test/core/test_type_utils.py

```diff
@@ -0,0 +1,326 @@
+        ),
+        pytest.param(
+            (Dict[(str, Mapping[(str, Dict[(str, int)])])]),
+            (Dict[(str, Mapping[(str, Dict[(Any, int)])])]),
+            id="nested-mapping-of-primitives-to-nested-mapping-of-any-keys",
+        ),
+        pytest.param(
+            (Dict[(str, Mapping[(str, Dict[(str, int)])])]),
+            (Dict[(str, Mapping[(Any, Dict[(str, int)])])]),
+            id="nested-mapping-of-primitives-to-nested-mapping-of-higher-level-any-keys",
+        ),
+        pytest.param(
+            (Dict[(str, Mapping[(str, Dict[(str, int)])])]),
+            (Dict[(str, Mapping[(str, Dict[(str, Any)])])]),
+            id="nested-mapping-of-primitives-to-nested-mapping-of-any-values",
+        ),
+        pytest.param(
+            (Dict[(str, Mapping[(str, Dict[(str, int)])])]),
+            (Dict[(str, Mapping[(Any, Dict[(Any, Any)])])]),
+            id="nested-mapping-of-primitives-to-nested-mapping-of-any-keys-and-values",
+        ),
+        pytest.param(
+            (Dict[(str, Mapping[(str, Dict[(str, Class1)])])]),
+            (Dict[(str, Mapping[(str, Dict[(str, Class1)])])]),
+            id="nested-mappings-of-same-classes",
+        ),
+        pytest.param(
+            (Dict[(str, Mapping[(str, Dict[(str, Class3)])])]),
+            (Dict[(str, Mapping[(str, Dict[(str, Class1)])])]),
+            id="nested-mapping-of-subclasses-to-nested-mapping-of-classes",
+        ),
+        pytest.param(
+            (Dict[(str, Mapping[(str, Dict[(str, Class1)])])]),
+            (Dict[(str, Mapping[(str, Dict[(Any, Class1)])])]),
+            id="nested-mapping-of-classes-to-nested-mapping-of-any-keys",
+        ),
+        pytest.param(
+            (Dict[(str, Mapping[(str, Dict[(str, Class1)])])]),
+            (Dict[(str, Mapping[(Any, Dict[(str, Class1)])])]),
+            id="nested-mapping-of-classes-to-nested-mapping-of-higher-level-any-keys",
+        ),
+        pytest.param(
+            (Dict[(str, Mapping[(str, Dict[(str, Class1)])])]),
+            (Dict[(str, Mapping[(str, Dict[(str, Any)])])]),
+            id="nested-mapping-of-classes-to-nested-mapping-of-any-values",
+        ),
+        pytest.param(
+            (Dict[(str, Mapping[(str, Dict[(str, Class1)])])]),
+            (Dict[(str, Mapping[(Any, Dict[(Any, Any)])])]),
+            id="nested-mapping-of-classes-to-nested-mapping-of-any-keys-and-values",
+        ),
+        pytest.param((Literal["a", "b", "c"]), (Literal["a", "b", "c"]), id="same-primitive-literal"),
+        pytest.param((Literal[Enum1.TEST1]), (Literal[Enum1.TEST1]), id="same-enum-literal"),
+        pytest.param(
+            (Tuple[(Optional[Literal["a", "b", "c"]], Union[(Path, Dict[(int, Class1)])])]),
+            (Tuple[(Optional[Literal["a", "b", "c"]], Union[(Path, Dict[(int, Class1)])])]),
+            id="identical-deeply-nested-complex-type",
+        ),
+    ],
+)
+def test_types_are_compatible(sender_type, receiver_type):
```

Comment:
This and the following test have been adapted from two similar tests in `test_connections.py`.